pkg/cloud/aws/actuators/machine/actuator.go

```diff
@@ -96,37 +96,50 @@ func machinesEqual(m1 *clusterv1.Machine, m2 *clusterv1.Machine) bool {
 	return m1.Name == m2.Name && m1.Namespace == m2.Namespace
 }
 
+// isNodeJoin determines if a machine, in scope, should join of the cluster.
 func (a *Actuator) isNodeJoin(scope *actuators.MachineScope, controlPlaneMachines []*clusterv1.Machine) (bool, error) {
 	switch set := scope.Machine.ObjectMeta.Labels["set"]; set {
 	case "node":
+		// Worker machines, not part of the controlplane, will always join the cluster.
 		return true, nil
 	case "controlplane":
+		// Controlplane machines will join the cluster if the cluster has an existing control plane.
+		controlplaneExists := false
+		var err error
 		for _, cm := range controlPlaneMachines {
 			m, err := actuators.NewMachineScope(actuators.MachineScopeParams{
-				Machine: cm,
-				Cluster: scope.Cluster,
-				Client:  a.clusterClient,
-				Logger:  a.log,
+				Machine:    cm,
+				Cluster:    scope.Cluster,
+				Client:     a.clusterClient,
+				Logger:     a.log,
+				AWSClients: scope.AWSClients,
 			})
 
 			if err != nil {
-				return false, errors.Wrapf(err, "failed to create machine scope for machine %q", cm.Name)
+				return false, errors.Wrapf(err, "failed to create machine scope for machine %q in namespace %q", cm.Name, cm.Namespace)
 			}
 
 			ec2svc := ec2.NewService(m.Scope)
 
-			ok, err := ec2svc.MachineExists(m)
+			controlplaneExists, err = ec2svc.MachineExists(m)
 			if err != nil {
-				return false, errors.Wrapf(err, "failed to verify existence of machine %q", m.Name())
+				return false, errors.Wrapf(err, "failed to verify existence of machine %q in namespace %q", m.Machine.Name, m.Machine.Namespace)
 			}
 
-			a.log.V(2).Info("Machine joining control plane", "machine-name", scope.Machine.Name, "machine-namespace", scope.Machine.Name, "should-join-control-plane", ok)
-			return ok, nil
+			if !controlplaneExists {
```

Comment:
This logic could be streamlined by moving log lines out of this function

```
for ... {
    controlPlaneExists, err := ec2svc.MachineExists(m)
    if controlPlaneExists {
        return controlPlaneExists, err
    }
}
return false, nil
```

This reduces the complexity of this code since the complexity comes from wanting to add a log line. If you feel the log line is necessary it could be pulled to the caller like this:

```
willJoin, err := isNodeJoin(...)
if willJoin {
     a.log.Info("node is joining the cluster")
}
```

Reply:
that would be better. Thanks!